### PR TITLE
angular-language-server: init at 18.2.0

### PR DESCRIPTION
**Pull request by tricktron**
EDIT:

Packaging the angular language server from source turns out to be tricky because it uses a combination of nodejs and bazel (and bazel is known to not play well together with nixpkgs https://github.com/NixOS/nixpkgs/issues/224446). The main problem is that bazel downloads a separate nodejs binary and then uses the pnpm lock file to transform all dependencies. Unfortunately, we cannot use that separate nodejs binary because it escapes the sandbox and probably will not even execute on nixos (e.g. nix-ld problem).

 I tried the following three approaches before I gave up and packged is as binary in https://github.com/NixOS/nixpkgs/pull/350257:

- https://github.com/NixOS/nixpkgs/pull/349521/commits/904f6667eb32381113bc007952659dc60f2dacce) was the start but I only realized afterwards that this only works without the sandbox on darwin.  On linux and darwin with the sandbox this fails because Bazel downloads its own nodejs binary which is not allowed in the buildPhase (no network access allowed).
- As a next step I tried to combine the yarn hooks with the buildBazelPackage function in https://github.com/NixOS/nixpkgs/pull/349521/commits/2e418c7d752e7e0938cc507599622e5b114c626f but that still failed because the downloaded nodejs binary still escapes the sandbox somehow. I then tried to manipulate the bazel files to inject the nix nodejs but then I got a bazel cycle dependency problem and since I am no bazel expert I stopped there.
- Finally, in https://github.com/NixOS/nixpkgs/pull/349521/commits/b1a395d84f0e1e6f99e3d2521ec9cdadbd8472ec I tried to replicate the esbuild bazel rules manually with the esbuild cli command but it still failed with some errors. I then gave up because this approach basically builds its own build script which will diverge in the future from upstream and does not seem maintainable to me. Besides you still need to be a bazel expert to continue transforming the upstream bazel build files to esbuild commands.

Continues on https://github.com/NixOS/nixpkgs/pull/343326.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/an/angular-language-server/package.nix [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+, gawk
+}:
+
+stdenvNoCC.mkDerivation (finalAttrs: {
+  pname = "angular-language-server";
+  version = "18.2.0";
+  src = fetchFromGitHub {
+    owner = "angular";
+    repo = "vscode-ng-language-service";
+    rev = "v${finalAttrs.version}";
```
Comment (NotAShelf) [inline pkgs/by-name/an/angular-language-server/package.nix:27]:
```suggestion
    rev = "refs/tags/v${finalAttrs.version}";
```

File: pkgs/by-name/an/angular-language-server/package.nix [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    typescript
+  ];
+
+  yarnBuildScript = "compile";
+
+  installPhase = ''
+    install -Dm755 bazel-bin/vsix_sandbox/server/bin/ngserver $out/bin/ngserver
+    install -Dm755 bazel-bin/vsix_sandbox/server/index.js $out/index.js
+    cp -r bazel-bin/vsix_sandbox/node_modules $out/node_modules
+  '';
```
Comment (NotAShelf) [inline pkgs/by-name/an/angular-language-server/package.nix:52]: Missing post/pre install hooks?

Comment (tricktron) [reply pkgs/by-name/an/angular-language-server/package.nix]: Jop, I forgot them, thank you for reminding me.

**Comment by tricktron** [comment]
Hm, I see that the build fails using the sandbox because during the buildPhase we switch from `yarn` to `bazel`. Since we don't have any internet access in the build phase this package becomes a lot more difficult to package. I don't have any experience with bazel, so any inputs are highly appreciated.

Maybe a double mkDerivation, e.g first derivation for the yarn deps to execute tsc -b and then a second mkDerivation for Bazel?

Pinging @lelgenio and @dotlambda for help.

**Comment by lelgenio** [comment]
I don't know much about bazel too, but I think you should be able to use `buildBazelPackage` in conjunction with the yarnHooks:

```nix
buildBazelPackage rec {
  pname = "angular-language-server";
  version = "18.2.0";

  src = fetchFromGitHub {
    owner = "angular";
    repo = "vscode-ng-language-service";
    rev = "refs/tags/v${version}";
    hash = "sha256-9+WWKvy/Vu4k0BzJwPEme+9+eDPI1QP0+Ds1CbErCN8=";
  };

  # yarn
  offlineCache = fetchYarnDeps {
    yarnLock = src + "/yarn.lock";
    hash = "sha256-N0N0XbNQRN7SHkilzo/xNlmn9U/T/WL5x8ttTqUmXl0=";
  };

  # bazel
  fetchAttrs.hash = "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=";

  nativeBuildInputs = [
    yarnConfigHook
    typescript
  ];

  bazel = bazel_5;

  buildAttrs.preBuild = ''
    tsc ....
  '';
  };
}
```

**Comment by lelgenio** [comment]
I haven't seen be used, and I'm not sure it would be acceptable, but a cursed alternative is to do the same thing as `node2nix`/`nodePackages`: Fetch the "pre-compiled" package from `npmjs.com` or `yarnpkg.com`

**Comment by SuperSandro2000** [comment]
If that works that would be awesome. We don't want to use node2nix, though.

File: pkgs/by-name/an/angular-language-server/package.nix [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    install -Dm755 server/bin/ngserver $out/bin/ngserver
+    install -Dm755 dist/index.js $out/index.js
+    cp -r node_modules $out/node_modules
+    runHook postInstall
+  '';
+
+  postFixup = ''
+    patchShebangs $out/bin/ngserver
+    patchShebangs $out/index.js
+    patchShebangs $out/node_modules
```
Comment (SuperSandro2000) [inline pkgs/by-name/an/angular-language-server/package.nix:88]:
```suggestion
    patchShebangs $out/bin/ngserver $out/index.js $out/node_modules
```

File: pkgs/by-name/an/angular-language-server/package.nix [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  #'';
+  buildPhase = ''
+        runHook preBuild
+        tsc -b server
+        ls -la dist/server/src
+        cat dist/server/src/banner.js
+        esbuild dist/server/src/server.js --bundle --banner:js="$(cat dist/server/src/banner.js)" --resolve-extensions=.js --outfile=dist/index.js --platform=node
+        ls -la
+        runHook postBuild
+    >>>>>>> 95143e16ffb5 (angular-language-server: manual build with esbuild)
```
Comment (SuperSandro2000) [inline pkgs/by-name/an/angular-language-server/package.nix:74]:
```suggestion
```

**Comment by tricktron** [comment]
@SuperSandro2000 @lelgenio @NotAShelf I gave up on packaging this from source and tried to document the encountered challenges  here https://github.com/NixOS/nixpkgs/pull/349521#issue-2597323661 for others who might start the same journey in the future.

In the meantime, I opened another pull request which packages it as a binary so that it is still availabe in nixpkgs over here: https://github.com/NixOS/nixpkgs/pull/350257.

What do you think?

**Comment by SuperSandro2000** [comment]
Fine for me, that's often the fate for bazel projects. Trash build system.

**Pull request closed without merging**